neothesia/src/scene/playing_scene/midi_player.rs

```diff
@@ -284,9 +470,89 @@ impl PlayAlong {
     }
 
     pub fn clear(&mut self) {
-        self.required_notes.clear()
+        self.required_notes.clear();
+        // Remove from the file log, notes that left pressed down with no key up yet (rewinding a non-played part)
+        self.file_notes.retain(|item| item.time_key_up.is_some());
+        self.user_pressed_recently.clear();
     }
 
+    pub fn finished(&mut self) {
+        if !self.finished {
+            // Loop through user_notes and file_notes and match entries with the same occurrence[note] = num
+            for user_note in &self.user_notes {
+                for file_note in &self.file_notes {
+                    if user_note.occurrence == file_note.occurrence
+                        && user_note.note_id == file_note.note_id
+                    {
+                        // Subtract timestamp from time_key_up to get total seconds
+                        let user_note_dur = match (user_note.timestamp, user_note.time_key_up) {
+                            (start, Some(end)) => end.duration_since(start).as_secs(),
+                            _ => 0,
+                        };
+                        let file_note_dur = match (file_note.timestamp, file_note.time_key_up) {
+                            (start, Some(end)) => end.duration_since(start).as_secs(),
+                            _ => 0,
+                        };
+
+                        // Add this information to user_stats.note_durations
+                        let note_duration = NoteDurations {
+                            user_note_dur: user_note_dur as usize,
+                            file_note_dur: file_note_dur as usize,
+                        };
+                        self.user_stats.note_durations.push(note_duration);
+                    }
+                }
+            }
+            //  Clear negative self.user_stats.wrong_notes
+            if self.user_stats.wrong_notes < 0 {
+                self.user_stats.wrong_notes = 0;
+            }
```

Comment:
This is fishy, how is it posible to have negative number of wrong notes? `wrong_notes` should be `u32` rather than `i32`

Reply:
You don't know it is a correct note until 500ms passed, so  I consider them all as wrong and subtract later if found to be correct, which may end up subtracting a zero catch by update().. validate that logic I am on the phone can't check out source rn

Reply:
We can use saturating_sub() if that's still the case with a u32, unless i stopped doing that subtracting in the later commits in user fn Because when a note pressed is removed from the pressed recently right there before reaching expiration, I am not sure may fn file still does the sub

Reply:
Yeha, I would rather just never let the value be invalid, rather than trying to fix the data afterwards. So saturating sub sounds good.